internals/idkey/idkey.go

```diff
@@ -0,0 +1,265 @@
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// idkey supplies an identity key for a machine, container or device. This
+// package provides an implementation based on an Ed25519 based key, currently
+// only supporting a file based key storage solution. This can later be
+// extended to support more secure hardware backed keystores, such as TPM,
+// OP-TEE or YubiKey.
+package idkey
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ed25519"
+	"crypto/rand"
+	"crypto/sha512"
+	"crypto/x509"
+	"encoding/base32"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+)
+
+const identityKeyFile = "key.pem"
+
+// Identity key must implement a crypto signer.
+var _ crypto.Signer = (*IDKey)(nil)
+
+type IDKey struct {
+	keyDir string
+	key    any
+}
+
+// New checks if an existing private identity key exists, and loads the key
+// if it does. It creates a new private identity key and persists it to disk
+// if no key was found. Cases where explicit control is desired on when to
+// generate or load can use GenerateKey and LoadKey directly.
+func New(keyDir string) (*IDKey, error) {
+	keyPath := filepath.Join(keyDir, identityKeyFile)
+	exists, err := pathExists(keyPath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot access key file %q: %w", keyPath, err)
+	}
+	if exists {
+		return LoadKey(keyDir)
+	}
+	return GenerateKey(keyDir)
+}
+
+// GenerateKey generates a new identity key and persists it to disk. This
+// function should only ever be called on the first boot otherwise the
+// existing identity will be overwritten.
+//
+// This function is equivalent to running:
+//
+//	openssl genpkey -algorithm Ed25519 -out key.pem
+func GenerateKey(keyDir string) (*IDKey, error) {
+	key := &IDKey{
+		keyDir: keyDir,
+	}
+	err := key.createDir()
+	if err != nil {
+		return nil, fmt.Errorf("cannot create identity directory: %w", err)
+	}
+	err = key.newEd25519()
+	if err != nil {
+		return nil, fmt.Errorf("cannot generate identity key: %w", err)
+	}
+	err = key.save()
+	if err != nil {
+		return nil, fmt.Errorf("cannot save identity key: %w", err)
+	}
+	return key, nil
+}
+
+// LoadKey loads an existing identity key from disk.
+func LoadKey(keyDir string) (*IDKey, error) {
+	key := &IDKey{
+		keyDir: keyDir,
+	}
+	// Load from disk.
+	err := key.load()
+	if err != nil {
+		return nil, fmt.Errorf("cannot load identity key: %w", err)
+	}
+	return key, nil
+}
+
+// createDir verifies the directory layout and permissions, and creates
+// the leaf element of the key driectory if it does not yet exist.
+func (k *IDKey) createDir() error {
+	exists, err := pathExists(k.keyDir)
+	if err != nil {
+		return err
+	}
+	if exists {
+		return expectPermission(k.keyDir, 0o700)
+	}
+
+	// Create the leaf directory node with 0o700 permissions.
+	return os.Mkdir(k.keyDir, 0o700)
+}
+
+// Public implements the crypto.Signer.
+func (k *IDKey) Public() crypto.PublicKey {
+	signer := k.key.(crypto.Signer)
```

Comment:
If `k.key` was typed as a `crypto.Signer`, we wouldn't need the type conversions at each access.

Reply:
Yes, after changing to ```ed25519.PrivateKey```, all these extra conversions fall away. Thanks. Resolving.